### PR TITLE
Fixes on the settlement related operations

### LINKED ISSUE #59: Update outdated documentations for the settlement functions
Some of the comments for the settlement functions are outdated after the refactors of the functions.

### LINKED ISSUE #60: Should re-throw exceptions whenever necessary for the settlement related functions.
As pointed out by @jonbern, it shouldn't totally silent the exception. Further thoughts, it does need to re-throw the unexpected exceptions, so easier for debugging for the downstream usages.

### LINKED ISSUE #62: Fix concurrent settlement operations
When settling or staging more than one settlement challenges in parallel for a wallet/currency pair, it will introduce race condition on ethereum nodes, resulting in transaction failures.

The solution to this is to ensure the transactions in the `startChallenge` or `settle` function of the `Settlement` class are executed in the sequence of transaction confirmations.

### LINKED ISSUE #65: Silent failure-and-continue
Swallowing error such as these can potentially lead to a disaster by allowing the function to keep running without handling the error. As a **minimum** the error should be debug logged. If its actually NOT an error, then the design should be reconsidered as it is using errors for something it shouldnt.

https://github.com/hubiinetwork/nahmii-sdk/blob/842d8cc97fe5717101b4b3ec256d25d8d355f612/lib/settlement.js#L56

### LINKED ISSUE #72: Undocumented public method
https://github.com/hubiinetwork/nahmii-sdk/blob/13b6010b6a488df1d591be1cccda46b5ad87659c/lib/wallet/wallet.js#L91

This method should be documented like the other methods. Does it really belong in the Wallet class though? Does it require a signature for anything? Maybe better located in the provider or a dedicated "BalanceTracker" class?

### PR DESCRIPTION
**Pull request by katat**
Closes #62 
Closes #59 
Closes #60 
Closes #65 
Closes #72

### CODE REVIEW

File: lib/settlement.js [hunk 1]
```diff
@@ -234,34 +234,74 @@ class Settlement {
...
     }
 
+    /**
+     * Start challenge based on the parameter object generated by #getRequiredChallengesForIntendedStageAmount.
+     * @param {RequiredChallenge} requiredChallenge - The object stores the necessary parameters for starting a settlement challenge.
+     * @param {Wallet} wallet - The wallet to start challenge
+     * @param [options]
+     * @returns {Promise} A promise that resolves into an transaction object.
+     */
+    async startByRequiredChallenge(requiredChallenge, wallet, options) {
```
[the remark this answers is not in the document]
Comment (katat) [reply lib/settlement.js]: Yes, they are used by CLI and core. Both CLI and core are calling this function to send and watch the transaction's status, in order to provide verbose messages when waiting for the transaction to be mined.

`startChallenge` and `settle` can be also used for the downstream for doing the same operations in an easier way, but the clients don't know what is going on until the all the transactions are mined or thrown errors.

File: lib/driip-settlement.js [hunk 2]
```diff
@@ -47,12 +49,15 @@ class DriipSettlement {
         try {
             return await challengeContract.proposalNonce(address, ct, id);
         } catch (err) {
-            return null;
+            if (isRevertContractException(err)) 
+                return null;
+            
+            throw err;
```
Comment (morfj) [inline lib/driip-settlement.js]: ´throw new Error('Unable to get current proposal nonce: ' + err);`

File: lib/driip-settlement.js [hunk 3]
```diff
@@ -174,17 +194,20 @@ class DriipSettlement {
         const settled = await this.hasPaymentDriipSettled(nonce, address);
         const latestNonce = await this.getCurrentProposalNonce(address, currency.ct, currency.id);
```
Comment (morfj) [inline lib/driip-settlement.js]: Run all in parallel using "Promise.all" ?

Comment (morfj) [reply lib/driip-settlement.js]: + catch and rethrow if anyone of these fails

Comment (katat) [reply lib/driip-settlement.js]: Updated using promise.all and rethrow using NestedError

File: lib/driip-settlement.js (hunk 2 above)
Comment (katat) [reply lib/driip-settlement.js]: Used NestedError class to do the rethrow. The code copied from the crawler repo.

**Comment by katat** [comment]
Hi @morfj, I have added the `NestedError` class to add more context information to the exceptions. The `NestedError` class does help a lot when it comes to debugging. Also the settlement related classes have been reorganized into the `/settlement` folder following the convention of the `/wallet` folder.

If these look good, I will release internal versions of the hubii core and cli for the team to play around with the withdrawal workflows, and see if there are any potential issues or making changes to the feedbacks.

Thanks

File: lib/utils.js [hunk 4]
```diff
@@ -250,6 +250,15 @@ function isSignedBy(message, signature, address) {
     }
 }
 
+/**
+ * Checks whether it is expected revert contract exception.
+ * @param {Error} error - error object thrown by ethers contract funcion call
+ * @returns {Boolean} True if it is revert exception thrown by the contract
+ */
+function isRevertContractException(error) {
```
Comment (morfj) [inline lib/utils.js]: Is this function really needed as an exported function from the SDK? Why should anyone outside the settlement implementations need to know about this type of error?

File: lib/settlement/index.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+'use strict';
+
+module.exports = require('./settlement');
```
Comment (morfj) [inline lib/settlement/index.js]: This file should also export the other public units from this module (driip-settlement and null-settlement) so users dont have to "deep require" files from inside a module directory.

**Comment by morfj** [comment]
This PR needs to also run the documentation updater (npm run build). The docs are slightly out of date now.

**Comment by morfj** [comment]
#72 could be fixed as part of this PR as well?

File: lib/settlement/index.js (hunk 5 above)
Comment (katat) [reply lib/settlement/index.js]: Good idea, I will update it.

File: lib/utils.js (hunk 4 above)
Comment (katat) [reply lib/utils.js]: You are right. For now, It is only used for the settlement operations. Perhaps I need to make it as a private function in the DriipSettlement and NullSettlement classes, but that could duplicate the code logic for that function. Another way is to create a `utils.js` inside the settlement folder to expose these types of functions.

**Comment by katat** [comment]
@morfj Yes, it will include #72

File: lib/utils.js (hunk 4 above)
Comment (morfj) [reply lib/utils.js]: Having a local "utils" inside the settlement folder sounds like a solution for now. I cant see that anyone outside "settlements" is using that function at present.

**Pull request merged**